### PR TITLE
Add the `on_before_backward` hook

### LINKED ISSUE #7789: Add training step callback hook before .backward
## 🚀 Feature
The proposal is to add a callback hook for the training step before .backward. Current hooks run either after .backward (on_after_backward) or after optimizer.step (on_train_batch_end, on_batch_end). The proposed callback hook would complement the existing hooks.

### Motivation

Adding this hook allows one to create callbacks related to loss calculations as well as batch operations for loss calculations. The current hooks cannot be used for this because the automatic pl optimizer is already called before the existing hooks. 

Adding the proposed hook would increase the flexibility and usability of pytorch lightning without modifying existing functionality. 

### Pitch

Concretely, it would suffice to add an additional callback hook, named for example "on_before_backward",  that takes in _trainer_ and _pl_module_ as arguments.

### Alternatives

Without the proposed callback, the alternative for the end user is to modify their own code to create their own hook in all the pl modules they use.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds `on_before_backward` hook and passes only the loss.

Fixes #7789

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
Is it not better to call `on_before_backward` in the training loop directly (instead of plugins) like we do for the `on_after_backward`? Or is there a blocker?

otherwise LGTM makes sense to include this hook

**Comment by tchaton** [comment]
> Is it not better to call `on_before_backward` in the training loop directly (instead of plugins) like we do for the `on_after_backward`? Or is there a blocker?
> 
> otherwise LGTM makes sense to include this hook

Hey @awaelchli. 

Calling `on_before_backward` in training_loop 
Pros: 
* Simpler and less code

Cons:
* When using precision, as a user, I would expect to receive the scaled loss.

IMO, I think this hook should be called right before backward and after any loss transformations.

Best,
T.C

**Comment by carmocca** [comment]
Should we also add it to `LightningModule`?

> The optimizer, optimizer_idx could also be passed along

`on_after_backward` doesn't include them so let's not include them here. We could do it in a future PR if somebody requests it.

**Comment by tchaton** [comment]
> Should we also add it to `LightningModule`?
> 
> > The optimizer, optimizer_idx could also be passed along
> 
> `on_after_backward` doesn't include them so let's not include them here. We could do it in a future PR if somebody requests it.

Yes, good idea !

File: tests/core/test_lightning_module.py [hunk 1]
```diff
@@ -379,3 +379,20 @@ def assert_device(device: torch.device) -> None:
     assert_device(torch.device("cpu"))
     trainer.predict(model, dataloaders=model.train_dataloader())
     assert_device(torch.device("cpu"))
+
+
+def test_on_before_backward(tmpdir):
```
Comment (carmocca) [inline tests/core/test_lightning_module.py:384]: Update this test instead: `test_trainer_model_hook_system_fit`

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 2]
```diff
@@ -74,6 +74,9 @@ def backward(
         # enter apex context
         closure_loss = scaled_loss.__enter__()
 
+        # hook
+        model.trainer.call_hook("on_before_backward", closure_loss)
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/apex_amp.py:78]: is there no way to avoid calling `model.trainer` ?

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1326,6 +1326,23 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer=None, opt_idx=None, *args, **kwargs)
         self._running_manual_backward = False
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
+        """
+        Override on_before_backward with your own implementation if you need to.
+
+        Args:
+            loss: Loss is already scaled by accumulated grads and possibly scaled by Mixed Precision.
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1334]: is it possible someone wants the unscaled loss?

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1326,6 +1326,23 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer=None, opt_idx=None, *args, **kwargs)
         self._running_manual_backward = False
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
+        """
+        Override on_before_backward with your own implementation if you need to.
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1331]: - it might not be clear to users that this is only called with automatic optimization. that also can lead to confusion over when to override this, optimizer step, or manual backward  and  when to set automatic_optimization=False.
- what's the guidance on this defined in the lightning module vs in core/hooks.py under ModelHooks?

i feel like if we could easily offer wrapped optimizers to users, would we still need all these intermediary hooks? would we feel more comfortable encouraging users to do manual optimization?

File: tests/core/test_lightning_module.py (hunk 1 above)
Comment (tchaton) [reply tests/core/test_lightning_module.py:384]: Done !

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:78]: I don't think there is.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1334]: I would assume the user to expect the final loss to be used for `backward` or open an issue if they want more.

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1331]: Hey @ananthsub,

For 1. This is being called independently of `automatic optimization`.
See below:
Screenshot 2021-06-08 at 11 05 44

For 2. For consistency reason, as ``on_after_backward`` is both a LightningModule / Callback hooks, it should be the same with ``on_before_backward``.

Not sure what you mean with `wrapped optimizers to users`. To me, this won't follow PyTorch API and I still prefer hooks through Callbacks and LightningModule.

``manual optimization`` is more about controlling the flow around ``backward`` and ``optimizer`` call. The hooks should still be called properly.

Adding hooks shoudn't make the ``manual optimization`` experience lower.

Best,
T.C

**Review by ethanwharris — APPROVED** [review]
LGTM

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1372,6 +1372,23 @@ def training_step(...):
         self.trainer.fit_loop.training_loop.batch_loop.backward(loss, optimizer=None, opt_idx=None, *args, **kwargs)
         self._running_manual_backward = False
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1375]: Move this to pytorch_lightning/core/hooks.py

**Review by carmocca — COMMENTED** [review]
Missing a call here

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/training_batch_loop.py#L570-L571

And a test for it

You can grep all the `on_after_backward` uses to check if you missed any

Also

- Update the hooks section of lightning_module.rst
- Add it to the API reference in lightning_module.rst
- Add it to the API reference in callbacks.rst

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/apex_amp.py:78]: yep, needs to call both the callbacks and the model hook

File: pytorch_lightning/core/lightning.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1376]: Done !

File: pytorch_lightning/core/hooks.py [hunk 6]
```diff
@@ -295,6 +295,23 @@ def on_before_zero_grad(self, optimizer: Optimizer) -> None:
             optimizer: The optimizer for which grads should be zeroed.
         """
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
+        """
+        Override on_before_backward with your own implementation if you need to.
+
+        Args:
+            loss: Loss is already scaled by accumulated grads and possibly scaled by Mixed Precision.
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:303]:
```suggestion
            loss: loss tensor (scalar), is already scaled by accumulated grads and possibly scaled by Mixed Precision.
```
scaled by accumulated grads? are you saying this is the accumulated loss or not?

File: pytorch_lightning/core/hooks.py [hunk 7]
```diff
@@ -295,6 +295,23 @@ def on_before_zero_grad(self, optimizer: Optimizer) -> None:
...
+
+        Called to before backward step.
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:305]:
```suggestion
        Called before ``loss.backward()``.
```

File: tests/plugins/test_deepspeed_plugin.py [hunk 8]
```diff
@@ -649,26 +650,40 @@ def test_deepspeed_multigpu_stage_2_accumulated_grad_batches(tmpdir, offload_opt
...
+            self.on_before_backward_called = False
+
         def on_train_batch_start(
             self, trainer, pl_module: LightningModule, batch: Any, batch_idx: int, dataloader_idx: int
         ) -> None:
             deepspeed_engine = trainer.training_type_plugin.model
             assert trainer.global_step == deepspeed_engine.global_steps
+            self.on_train_batch_start_called = True
+
+        def on_before_backward(self, trainer, pl_module, loss):
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:663]: is this something that you specifically need for deepspeed?

**Comment by tchaton** [comment]
> Missing a call here
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/training_batch_loop.py#L570-L571
> 
> And a test for it
> 
> You can grep all the `on_after_backward` uses to check if you missed any
> 
> Also
> 
> * Update the hooks section of lightning_module.rst
> * Add it to the API reference in lightning_module.rst
> * Add it to the API reference in callbacks.rst

Hey @carmocca,

The call is made inside the precision plugin, so we can capture the loss which is actually used for training.
And there is a test for all `deepseed plugin` +  `normal run`.

Best,
T.C

**Comment by carmocca** [comment]
Also maybe pass batch idx just like for after bwd?

**Review by awaelchli — COMMENTED** [review]
Overall looks pretty smooth. 

Note, there is a file notebooks submodule that your git checkout accidentally deleted. Happened to me too, it's weird ...

File: tests/plugins/test_deepspeed_plugin.py (hunk 8 above)
Comment (tchaton) [reply tests/plugins/test_deepspeed_plugin.py:663]: No, just a test to make sure it is being called properly.

File: pytorch_lightning/core/hooks.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:303]: Reformulated.

File: pytorch_lightning/core/hooks.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:305]: Done !

File: pytorch_lightning/trainer/callback_hook.py [hunk 9]
```diff
@@ -313,6 +315,13 @@ def on_load_checkpoint(self, checkpoint):
                 else:
                     callback.on_load_checkpoint(self, self.lightning_module, state)
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
+        """
+        Called before loss.backward().
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:320]: Done !

File: pytorch_lightning/core/hooks.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:303]: Done !

File: pytorch_lightning/core/hooks.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:305]: Done !

**Review by tchaton — COMMENTED** [review]
> Also maybe pass batch idx just like for after bwd?

```
    def on_before_backward(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', loss: torch.Tensor) -> None:
        """Called before ``loss.backward()``."""
        pass

    def on_after_backward(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule') -> None:
        """Called after ``loss.backward()`` and before optimizers do anything."""
        pass
```

@carmocca We don't pass batch_idx. Might be a good idea though.

File: tests/plugins/test_deepspeed_plugin.py (hunk 8 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:663]: @tchaton why is this in the deepspeed test? Should it not be in it's own general test?

**Review by SeanNaren — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: tests/plugins/test_amp_plugins.py [hunk 10]
```diff
@@ -69,18 +70,34 @@ def test_amp_apex_ddp(
     assert isinstance(trainer.precision_plugin, plugin_cls)
 
 
+class CheckOnBeforeBackward(Callback):
```
Comment (carmocca) [inline tests/plugins/test_amp_plugins.py:73]: what if instead of adding these callbacks per plugin test you parameterize the test in `test_hooks`?

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 11]
```diff
@@ -59,6 +59,10 @@ def backward(
         """
         closure_loss = self.scaler.scale(closure_loss)
 
+        # hook
+        model.trainer.call_hook("on_before_backward", closure_loss)
+
+        # call `on_before_backward` hook
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/native_amp.py:65]: twice here?

Comment (tchaton) [reply pytorch_lightning/plugins/precision/native_amp.py:65]: removed !

File: tests/plugins/test_amp_plugins.py (hunk 10 above)
Comment (tchaton) [reply tests/plugins/test_amp_plugins.py:73]: Hey @carmocca,
I prefer the current approach as testing `on_before_backward` for all plugins will result in a large increase of CI time.
I think it is light enough to be embedded into existing tests.

Comment (carmocca) [reply tests/plugins/test_amp_plugins.py:73]: Note that we also get the advantage of checking all other hooks, not just `on_before_backward`.

The hook order should be consistent among accelerators/plugins/training_types so we eventually want to test it

**Review by carmocca — APPROVED** [review]
